File: skills/openclix-init/templates/ios/Engine/ScheduleCalculator.swift
import Foundation

public struct ScheduleInput {
    public let now: String
    public let execute_at: String?
    public let delay_seconds: Int?
    public let do_not_disturb: DoNotDisturb?

    public init(
        now: String,
        execute_at: String? = nil,
        delay_seconds: Int? = nil,
        do_not_disturb: DoNotDisturb? = nil
    ) {
        self.now = now
        self.execute_at = execute_at
        self.delay_seconds = delay_seconds
        self.do_not_disturb = do_not_disturb
    }
}

public struct ScheduleResult {
    public let execute_at: String
    public let skipped: Bool
    public let skip_reason: SkipReason?

    public init(execute_at: String, skipped: Bool, skip_reason: SkipReason? = nil) {
        self.execute_at = execute_at
        self.skipped = skipped
        self.skip_reason = skip_reason
    }
}

private func isInDoNotDisturbWindow(hour: Int, doNotDisturb: DoNotDisturb) -> Bool {
    let startHour = doNotDisturb.start_hour
    let endHour = doNotDisturb.end_hour

    if startHour <= endHour {
        return hour >= startHour && hour < endHour
    }

    return hour >= startHour || hour < endHour
}

public final class ScheduleCalculator {

    public init() {}

    public func calculate(_ input: ScheduleInput) -> ScheduleResult {
        var executeAtDate: Date

        if let executeAt = input.execute_at,
           let parsedDate = ISO8601DateFormatter().date(from: executeAt) {
            executeAtDate = parsedDate
        } else if let parsedDate = ISO8601DateFormatter().date(from: input.now) {
            executeAtDate = parsedDate
            if input.execute_at == nil,
               let delaySeconds = input.delay_seconds,
               delaySeconds > 0 {
                executeAtDate = parsedDate.addingTimeInterval(Double(delaySeconds))
            }
        } else {
            executeAtDate = Date()
        }

        if let doNotDisturb = input.do_not_disturb {
            let hour = Calendar.current.component(.hour, from: executeAtDate)
            if isInDoNotDisturbWindow(hour: hour, doNotDisturb: doNotDisturb) {
                return ScheduleResult(
                    execute_at: executeAtDate.formattedISO8601(),
                    skipped: true,
                    skip_reason: .campaign_do_not_disturb_blocked
                )
            }
        }

        return ScheduleResult(
            execute_at: executeAtDate.formattedISO8601(),
            skipped: false
        )
    }
}

private extension Date {
    func formattedISO8601() -> String {
        return ISO8601DateFormatter().string(from: self)
    }
}

File: skills/openclix-init/templates/android/engine/ScheduleCalculator.kt
package ai.openclix.engine

import ai.openclix.models.DoNotDisturb
import ai.openclix.models.SkipReason
import java.util.Calendar
import java.util.Date

// Input supports either a precomputed execute_at or now + delay_seconds.
data class ScheduleInput(
    val now: String,
    val execute_at: String? = null,
    val delay_seconds: Int? = null,
    val do_not_disturb: DoNotDisturb? = null
)

data class ScheduleResult(
    val execute_at: String,
    val skipped: Boolean,
    val skip_reason: SkipReason? = null
)

private fun isInDoNotDisturbWindow(hour: Int, doNotDisturb: DoNotDisturb): Boolean {
    val startHour = doNotDisturb.start_hour
    val endHour = doNotDisturb.end_hour
    return if (startHour <= endHour) {
        hour >= startHour && hour < endHour
    } else {
        hour >= startHour || hour < endHour
    }
}

class ScheduleCalculator {

    fun calculate(input: ScheduleInput): ScheduleResult {
        val nowDate = Date(parseIso8601(input.now))
        var executeAtDate = if (input.execute_at != null) {
            Date(parseIso8601(input.execute_at))
        } else {
            nowDate
        }

        val executeAtMillis = executeAtDate.time
        if (executeAtMillis <= 0L) {
            executeAtDate = nowDate
        } else if (input.execute_at == null && input.delay_seconds != null && input.delay_seconds > 0) {
            executeAtDate = Date(executeAtDate.time + input.delay_seconds * 1000L)
        }

        if (input.do_not_disturb != null) {
            val calendar = Calendar.getInstance().apply { time = executeAtDate }
            val hour = calendar.get(Calendar.HOUR_OF_DAY)
            if (isInDoNotDisturbWindow(hour, input.do_not_disturb)) {
                return ScheduleResult(
                    execute_at = toIso8601(executeAtDate.time),
                    skipped = true,
                    skip_reason = SkipReason.CAMPAIGN_DO_NOT_DISTURB_BLOCKED
                )
            }
        }

        return ScheduleResult(
            execute_at = toIso8601(executeAtDate.time),
            skipped = false
        )
    }
}

internal fun parseIso8601(isoString: String): Long {
    return try {
        java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.US).apply {
            timeZone = java.util.TimeZone.getTimeZone("UTC")
        }.parse(isoString)?.time ?: System.currentTimeMillis()
    } catch (_: Exception) {
        try {
            java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.US).apply {
                timeZone = java.util.TimeZone.getTimeZone("UTC")
            }.parse(isoString)?.time ?: System.currentTimeMillis()
        } catch (_: Exception) {
            System.currentTimeMillis()
        }
    }
}

internal fun toIso8601(epochMilliseconds: Long): String {
    val formatter = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.US)
    formatter.timeZone = java.util.TimeZone.getTimeZone("UTC")
    return formatter.format(java.util.Date(epochMilliseconds))
}
